File: Data Structures & Algorithms/island-perimeter/submission-0.java
class Solution {
    public int dfs(int i, int j, int[][] grid, boolean[][] visited) {
        if (i >= grid.length || i < 0 || j >= grid[0].length || j < 0 
            || grid[i][j] == 0) {
            return 1;
        }

        if (visited[i][j]) {
            return 0;
        }

        visited[i][j] = true;

        int right = dfs(i, j + 1, grid, visited);
        int left = dfs(i, j - 1, grid, visited);
        int up = dfs(i + 1, j, grid, visited);
        int down = dfs(i - 1, j, grid, visited);


        return right + left + up + down;
    }
    public int islandPerimeter(int[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[r].length; c++) {
                visited[r][c] = false;
            }
        }

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[r].length; c++) {
                if (grid[r][c] == 1) {
                    return dfs(r, c, grid, visited);
                }
            }
        }

        return dfs(0, 0, grid, visited);
    }
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        ROWS = len(matrix)
        COLS = len(matrix[0])

        memo = {}
        def dfs(r, c, prev_val):
            if r < 0 or r >= ROWS or c < 0 or c >= COLS or matrix[r][c] <= prev_val:
                return 0
            
            if (r, c) in memo:
                return memo[(r, c)]

            down = 1 + dfs(r + 1, c, matrix[r][c])
            up = 1 + dfs(r - 1, c, matrix[r][c])
            right = 1 + dfs(r, c + 1,matrix[r][c])
            left = 1 + dfs(r, c - 1, matrix[r][c])

            memo[(r, c)] = max(down, up, right, left)
            return memo[(r, c)]

        maxx = 0
        for r in range(ROWS):
            for c in range(COLS):
                maxx = max(dfs(r, c, -1), maxx)
        
        return maxx